Source/ARTRealtime.m

```diff
@@ -849,28 +849,22 @@ - (void)onHeartbeat {
 - (void)onConnected:(ARTProtocolMessage *)message {
     _renewingToken = false;
     
-    // Resuming
-    if (_resuming) {
-        if (![message.connectionId isEqualToString:self.connection.id_nosync]) { // RTN15c3
-            ARTLogWarn(self.logger, @"RT:%p connection \"%@\" has reconnected, but resume failed. Reattaching any attached channels", self, message.connectionId);
-            // Reattach all channels
-            for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
-                ARTAttachRequestMetadata *const metadata = [[ARTAttachRequestMetadata alloc] initWithReason:message.error];
-                [channel reattachWithMetadata:metadata];
-            }
-            _resuming = false;
-        }
-        else if (message.error) {
-            ARTLogWarn(self.logger, @"RT:%p connection \"%@\" has resumed with non-fatal error \"%@\"", self, message.connectionId, message.error.message);
-            // The error will be emitted on `transition`
-        }
-        else {
-            ARTLogDebug(self.logger, @"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId);
-        }
-    }
-    
     switch (self.connection.state_nosync) {
         case ARTRealtimeConnecting: {
+            if (_resuming) {
+                if ([message.connectionId isEqualToString:self.connection.id_nosync]) {
+                    ARTLogDebug(self.logger, @"RT:%p connection \"%@\" has reconnected and resumed successfully", self, message.connectionId);
+                }
+                else {
+                    ARTLogWarn(self.logger, @"RT:%p connection \"%@\" has reconnected, but resume failed. Error: \"%@\"", self, message.connectionId, message.error.message);
+                }
+                // Reattach all channels regardless resume success - RTN15c6, RTN15c7
+                for (ARTRealtimeChannelInternal *channel in self.channels.nosyncIterable) {
```

Comment:
This check is inside `[channel reattachWithMetadata:metadata]` call. Probably it's not the best idea tho.

Reply:
Ah, OK. How about the tests? I guess that the existing check in `reattachWithMetadata` existed because of RTN15c3, which has now been removed along with its test.

Reply:
1af28a44aded9df8055c91178845b2e169659078 ab6c6d34855f48b9d81ebfe2649c250d78251426 now checks for detaching channels as well
I'm not sure about SUSPENDED channels in RTN15c6, because after channel's got suspended there should not be successful resume. I'm getting different connection id with no error after channel became suspended - https://github.com/ably/ably-cocoa/pull/1714#discussion_r1299426638

Reply:
> because after channel's got suspended there should not be successful resume

What makes you believe that? Do you think there's something wrong with the spec?

Reply:
Looks like specs don't explicitly state this except these two hints: [RTN15g](https://sdk.ably.com/builds/ably/specification/main/features/#RTN15g) and [RTL3c](https://sdk.ably.com/builds/ably/specification/main/features/#RTL3c). But [docs do](https://ably.com/docs/channels?lang=swift#non-fatal-errors):

> 
> - If a client [resumes or recovers](https://ably.com/docs/connect/states) a connection unsuccessfully continuity is lost and the client receives a fresh connection. This generally happens because the client was disconnected for more than two minutes, which is how long Ably holds connection state for. If the client were resuming, all the channels (which will have gone into the SUSPENDED state after two minutes) will still reattach automatically, and the client will receive ATTACHED events with resumed set to false.
> - If Ably needs to signal a loss of message continuity on an attached channel, the client will receive an UPDATE event with resumed set to false. This occurs in situations such as a partially successful resume, where the client was disconnected for less than two minutes.

Reply:
Yeah, I think that the main reason for a channel to enter the `SUSPENDED` state is due to the connection entering the `SUSPENDED` state (RTL3c), which implies loss of connection state.

But the spec does describe some other ways in which a channel can become `SUSPENDED`; e.g. [RTN4f](https://sdk.ably.com/builds/ably/specification/main/features/#RTL4f) or [RTN13b](https://sdk.ably.com/builds/ably/specification/main/features/#RTL13b). These ones do not necessarily imply a loss of connection state. Perhaps we could trigger one of these scenarios to put the channel into the `SUSPENDED` state, and then check that this channel gets re-attached upon a RTN15c6 or RTN15c7 resume?

Reply:
6af5337ef4507987a48bed575aba7a53b236c5d8